File: robot/firmware/eeprom.c
/************************************************************************ 

    eeprom.c

    Valiant Turtle 2 - Raspberry Pi Pico W Firmware
    Copyright (C) 2024 Simon Inns

    This file is part of Valiant Turtle 2

    This is free software: you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Email: simon.inns@gmail.com

************************************************************************/

#include <stdio.h>
#include <stdlib.h>

#include "pico/stdlib.h"
#include "hardware/i2c.h"

#include "i2cbus.h"
#include "eeprom.h"
#include "debug.h"

// Initialise the 24LC16 I2C EEPROM
void eeprom_initialise()
{
    // Nothing needs to be done here
}

// Read a byte from the 24LC16 I2C EEPROM
uint8_t eeprom_read_byte(uint16_t address)
{
    // Ensure the address is not out of bounds
    if ((address) > (256*8)) {
        debug_printf("eeprom_read_byte(): Requested address is out of bounds!\n");
        return 0;
    }

    // Determine the device address (including the block) and intra-block address
    uint8_t device_address = EEPROM_ADDR | (address >> 8);
    uint8_t block_address = (address & 0xFF);

    // Perform the read by sending the block, block address and then reading
    // one byte from the device
    uint8_t data;
    i2c_write_blocking(i2c0, device_address, &block_address, 1, true); 
    i2c_read_blocking(i2c0, device_address, &data, 1, false);

    return data;
}

void eeprom_write_byte(uint16_t address, uint8_t data)
{
    // Ensure the address is not out of bounds
    if ((address) > (256*8)) {
        debug_printf("eeprom_write_byte(): Requested address is out of bounds!\n");
        return;
    }

    // Determine the device address (including the block) and intra-block address
    uint8_t device_address = EEPROM_ADDR | (address >> 8);
    uint8_t block_address = (address & 0xFF);

    uint8_t command[2];
    command[0] = block_address;
    command[1] = data;

    i2c_write_blocking(i2c0, device_address, command, 2, false);
    sleep_us(11000);
}

// Read one or more bytes from the 24LC16 I2C EEPROM
void eeprom_read(uint16_t address, uint8_t *data, uint16_t data_length)
{
    // Ensure the address is not out of bounds
    if ((address+data_length) > (256*8)) {
        debug_printf("eeprom_read(): Requested address is out of bounds!\n");
        return;
    }

    // Determine the device address (including the block) and intra-block address
    uint8_t device_address = EEPROM_ADDR | (address >> 8);
    uint8_t block_address = (address & 0xFF);

    // Write the required starting address
    i2c_write_blocking(i2c0, device_address, &block_address, 1, true); 

    // Now read the required data (the whole EEPROM can be read in one command)
    i2c_read_blocking(i2c0, device_address, data, (size_t)data_length, false);
}

void eeprom_write(uint16_t address, uint8_t *data, uint16_t data_length)
{
    // Ensure the address is not out of bounds
    if ((address+data_length) > (256*8)) {
        debug_printf("eeprom_write(): Requested address is out of bounds!\n");
        return;
    }

    // Note: The page size of 16 bytes is the maximum we can write in one command...
    int32_t remaining_data = data_length;
    int32_t data_pointer = 0;

    while(remaining_data > 0) {
        // Determine the start address for this page write
        uint8_t device_address = EEPROM_ADDR | ((address+data_pointer) >> 8);
        uint8_t block_address = ((address+data_pointer) & 0xFF);

        // Initialise an array to store the start address and a page of data
        uint8_t page[16+1];
        int32_t write_length = 16;

        // If we have less that 16 bytes of remaining data, lower the write length
        if (remaining_data < 16) write_length = remaining_data;

        // If we would cross a block boundary, lower the write length to avoid
        if ((block_address+write_length) > 255) write_length -= (block_address+write_length) - 256;

        // Copy the start address and the data into the page
        page[0] = block_address;
        for (int32_t i=0; i < write_length; i++) {
            page[i+1] = data[data_pointer];
            data_pointer++;
        }

        // Perform a page write
        i2c_write_blocking(i2c0, device_address, page, write_length+1, false);
        sleep_us(11000);

        // Continue if required
        remaining_data -= write_length;
    }
}

File: robot/firmware/eeprom.py
#************************************************************************ 
#
#   eeprom.py
#
#   24LC16 I2C EEPROM Communication
#   Valiant Turtle 2 - Robot firmware
#   Copyright (C) 2024 Simon Inns
#
#   This file is part of Valiant Turtle 2
#
#   This is free software: you can redistribute it and/or
#   modify it under the terms of the GNU General Public License as
#   published by the Free Software Foundation, either version 3 of the
#   License, or (at your option) any later version.
#
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#   Email: simon.inns@gmail.com
#
#************************************************************************

import logging
from machine import I2C
from time import sleep

class Eeprom:
    """
    A class to represent a 24LC16 I2C EEPROM.

    Attributes
    ----------
    i2c : I2C
        The I2C bus instance.
    i2c_address : int
        The I2C address of the EEPROM.
    _is_present : bool
        Flag indicating if the EEPROM is present.
    _maximum_address : int
        The maximum address value for the 24LC16 EEPROM.

    Methods
    -------
    __init__(i2c_bus: I2C, address: int = 0x50):
        Initializes the EEPROM with the given I2C bus and address.
    """

    def __init__(self, i2c_bus: I2C, address: int = 0x50):
        """
        Constructs all the necessary attributes for the EEPROM object.

        Parameters
        ----------
        i2c_bus : I2C
            The I2C bus instance.
        address : int, optional
            The I2C address of the EEPROM (default is 0x50).
        """
        self.i2c = i2c_bus
        self.i2c_address = address

        # Check that the EEPROM is present
        self._is_present = False

        devices = self.i2c.scan()
        for idx in range(len(devices)):
            if devices[idx] == self.i2c_address: self._is_present = True

        if self._is_present:
            logging.info(f"Eeprom::__init__ - 24LC16 EEPROM detected at address {hex(self.i2c_address)}")
        else:
            logging.info("Eeprom::__init__ - 24LC16 EEPROM is not present... Cannot initialise!")

        self._maximum_address = 2048 # Maximum address value for the 24LC16
        self._page_size = 16 # The page size for the 24LC16

    # Return True is 24LC16 EEPROM was detected and initialised
    @property
    def is_present(self):
        return self._is_present
    
    def read(self, address, number_of_bytes) -> bytes:
        # Ensure the whole read is within the memory's boundaries
        if address + number_of_bytes > self._maximum_address:
            raise ValueError("Eeprom::read - Requested address exceeds the maximum allowed")
        if address < 0:
            raise ValueError("Eeprom::read - Requested address is less than zero")
        
        # Determine the device address (including the block) and intra-block address
        self._devaddr = bytearray(1)
        self._blockaddr = bytearray(1)
        self._devaddr[0] = self.i2c_address | (address >> 8) # Device address
        self._blockaddr[0] = (address & 0xFF); # Block address

        # Read from the EEPROM and return the collected data
        logging.debug(f"Eeprom::read - Address = {address} - number of bytes = {number_of_bytes}")

        sleep(0.01)
        self.i2c.writeto(self._devaddr[0], self._blockaddr, False)
        return self.i2c.readfrom(self._devaddr[0], number_of_bytes, True)
    
    def write(self, address, data: bytes):
        # Ensure the whole write is within the memory's boundaries
        if address + len(data) > self._maximum_address:
            raise ValueError("Eeprom::write - Requested address exceeds the maximum allowed")
        if address < 0:
            raise ValueError("Eeprom::write - Requested address is less than zero")
        
        remaining_data = len(data) # Keep track of what's left to write
        data_pointer = 0
        logging.debug(f"Eeprom::write - Writing address = {address} - write length = {remaining_data}")

        while remaining_data > 0:
            # Determine the device address (including the block) and intra-block address
            self._addr = bytearray(2)
            self._addr[0] = self.i2c_address | (address >> 8) # Device address
            self._addr[1] = (address & 0xFF) # Block address

            # Maximum write length is the page size
            write_length = self._page_size
            if (remaining_data < self._page_size): 
                write_length = remaining_data

            # If the required read would cross and block boundary, lower the 
            # write length to prevent this happening
            bytes_until_boundary = self._page_size - address % self._page_size
            if (write_length > bytes_until_boundary): write_length = bytes_until_boundary

            # Copy the start address and the data into the page buffer
            page_buffer = bytearray(write_length+1)
            page_buffer[0] = self._addr[1]
            for i in range(0, write_length):
                page_buffer[i+1] = data[data_pointer]
                data_pointer += 1

            logging.debug(f"Eeprom::write - Page write @ address = {address} - write length = {write_length}")

            # Perform a write
            sleep(0.01)
            self.i2c.writeto(self._addr[0], page_buffer, True)

            # Continue if required
            remaining_data -= write_length
            address += write_length

if __name__ == "__main__":
    from main import main
    main()